### PR TITLE
Improve Docker Support

### PR DESCRIPTION
**Pull request by slmg**
Dockerfile

* Deterministic build
* Run as user `node` instead of `root`
* Various env to make the app happier

docker-compose.yml

* Use a volume to preserve state of game data

Other
* Add doc in Readme
* Stop sending *node_modules/* context to the docker engine
* Fix path in *routes/index.js*

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
-COPY ["src", "package.json", "/app/"]
+WORKDIR ${APP_DIR}
 
-RUN npm install --production \
-  && mkdir -p ${APP_DIR}/saved
+COPY src/ package*.json ./
 
-CMD ["node", "server.js"]
\ No newline at end of file
+RUN npm install && \
```
[the remark this answers is not in the document]
Comment (slmg) [reply Dockerfile:10]: It's not needed due to the declaration of `NODE_ENV` a bit before.
From <https://docs.npmjs.com/cli/install>:
>With the --production flag (or when the NODE_ENV environment variable is set to production), npm will not install modules listed in devDependencies.

File: Dockerfile [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
+    mkdir saved && \
+    chown -R node:node ${APP_DIR}
+
+USER node
+
+ENV HOST localhost
```
[the remark this answers is not in the document]
Comment (slmg) [reply Dockerfile:16]: I can if you wish, they were intentionally put at the end for this reason: <https://runnable.com/blog/9-common-dockerfile-mistakes>
>EXPOSE and ENV are cheap commands to run. If you bust the cache for them, rebuilding them is almost instantaneous. Therefore, it’s best to declare these commands as late as possible. You should only ever declare ENVs whenever you need them in your build process. If they’re not needed during build time, then they should be at the end of your Dockerfile, along with EXPOSE.

File: docker-compose.yml [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
     build:
       context: .
-      args:
-        APP_DIR: ${APP_DIR}
     image: vim-adventures:latest
+    container_name: vim-adventures
     ports:
-     - ${PORT}:${PORT}
-    env_file: .env
+      - "${PORT:-127.0.0.1:8080}:8080"
```
[the remark this answers is not in the document]
Comment (slmg) [reply docker-compose.yml:10]: I have detailed this in the Readme.

File: README.md [hunk 4]
```diff
@@ -10,59 +11,84 @@ This is a great game and if you do use it, please head to the [vim adventures](h
...
+Build
+
+```sh
+docker build -t vim-adventures:latest .
+```
+
+Run
+
+```sh
+docker container run -d \
```
[the remark this answers is not in the document]
Comment (slmg) [reply README.md:90]: Ok I can remove the `-d` parameter. Alternatively I can leave it and add a `--name vim-adventures` and explain in the Readme that stop/start can be done using `docker container stop vim-adventures` and logs accessed using `docker container logs [-f] vim-adventures`. What do you think?

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (09wattry) [reply Dockerfile:10]: If that is the case we should also allow that variable to be set at build time

[the remark this answers is not in the document]
Comment (slmg) [reply Dockerfile:10]: I am not sure I understand what you mean?

File: Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (09wattry) [reply Dockerfile:16]: That makes sense, you are probably more versed in docker than I. Thanks for the resource.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (09wattry) [reply README.md:90]: I don't mind, the later is just more work for you so I'll leave it up to you to decide.

File: README.md [hunk 5]
```diff
@@ -10,59 +11,84 @@ This is a great game and if you do use it, please head to the [vim adventures](h
...
-* ```APP_NAME``` - name you would like to export for the application. Default - vim-adventures
-* ```PORT``` - the port that you want the application to run on localhost. - Default - 80
+#### Using Docker Compose
 
-This .env file will be copied to the install install directory i.e. APP_DIR and can be updated from there after installation.
+Simply run:
 
-## <span id="run">Running Vim Adventures</span>
+```sh
+docker-compose up
```
Comment (09wattry) [inline README.md:66]: They will have to run a docker-compose build first if they do not have the image first right?

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (slmg) [reply README.md:90]: Ok I'll remove the `-d` param :smile:

File: docker-compose.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (09wattry) [reply docker-compose.yml:10]: I see that now, I was unsure if it would override what was in the .env file.

File: README.md (hunk 5 above)
Comment (slmg) [reply README.md:66]: No `docker-compose up` nicely takes care of building the image beforehand (only if `build` is defined in the yaml file)

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (09wattry) [reply Dockerfile:10]: You know what at the end of the day we will only be releasing production-ready images so a hardcoded production env is not a problem right now.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (slmg) [reply README.md]: I also added `--rm` to automatically remove stopped containers and `--pid=host` due to a bug preventing killing the server by using Ctrl+C.
<https://github.com/nodejs/docker-node/issues/848>
<https://stackoverflow.com/questions/36637912/how-to-stop-running-node-in-docker>

**Comment by 09wattry** [comment]
@slmg so I'm sorry I had to do this but I needed to do a whole lot of work to get the master branch back to a decent state. I don't like multi commit pulls unless they are necessary so I rehauled the master branch.

I'm not sure how your git workflow skills are but you're gonna need to do a ```git pull --rebase``` against my master branch and throw out all of my old commits. Once it's all clean push up your new branch and we can merge your changes in. If you need some help let me know and we can try work it out together.

Again, sorry to do this to you.

**Comment by slmg** [comment]
I think the branch divergence is resolved. I rebased my local branch on the latest common commit before force-pushing it.
I think it may be possible to test your CI/CD from a branch, it could prevent this situation to happen again. The master branch is supposed to be sacred and never have its history rewritten :wink:

**Comment by 09wattry** [comment]
> I think the branch divergence is resolved. I rebased my local branch on the latest common commit before force-pushing it.
> I think it may be possible to test your CI/CD from a branch, it could prevent this situation to happen again. The master branch is supposed to be sacred and never have its history rewritten 😉

I know, I know just don't tell anyone okay ;) The CICD is working, for the most part, its creating binaries so that people can just download the binary and play if they don't want to clone the repo or install Node or Docker.

**Comment by slmg** [comment]
Nice! :tada: I let you merge I am not permissioned.

**Pull request merged**